abel/benchmark.py

```diff
@@ -14,199 +14,504 @@
+    """
+    return x if isinstance(x, list) else [x]
+
+
+def _roundsf(x, n):
+    """
+    Round to n significant digits
+    """
+    return float('{:.{p}g}'.format(x, p=n))
+
+
+class Timent(object):
+    """
+    Helper class for measuring execution times.
+
+    The constructor only initializes the timing-procedure parameters.
+    Use the :py:meth:`.time` method to run it for particular functions.
+
+    Parameters
+    ----------
+    skip : int
+        number of "warm-up" iterations to perform before the measurements.
+        Can be specified as a negative number, then ``abs(skip)``
+        "warm-up" iterations are performed, but if this took more than
+        **duration** seconds, they are accounted towards the measured
+        iterations.
+    repeat : int
+        minimal number of measured iterations to perform.
+        Must be positive.
+    duration : float
+        minimal duration (in seconds) of the measurements.
+    """
+    def __init__(self, skip=0, repeat=1, duration=0.0):
+        self.skip = int(skip)
+        self.repeat = int(repeat)
+        self.duration = float(duration)
+
+    def time(self, func, *args, **kwargs):
         """
-        Benchmark performance of different iAbel/fAbel implementations.
+        Repeatedly executes a function at least **repeat** times and for at
+        least **duration** seconds (see above), then returns the average time
+        per iteration.
+        The actual number of measured iterations can be retrieved from
+        :py:attr:`Timent.count`.
 
         Parameters
         ----------
-        n: integer
-            a list of arrays sizes for the benchmark
-            (assuming 2D square arrays (n,n))
-
-        select: list of str
-            list of transforms to benchmark select=['all',] (default) or 
-            choose transforms:
-            select=['basex', 'direct_Python', 'direct_C', 'hansenlaw', 
-                    'linbasex' 'onion_bordas, 'onion_peeling', 'two_point', 
-                    'three_point']
+        func : callable
+            function to execute
+        *args, **kwargs : any, optional
+            parameters to pass to **func**
 
-        n_max_bs: integer
-            since the basis sets generation takes a long time,
-            do not run this benchmark for implementations that use basis sets
-            for n > n_max_bs
+        Returns
+        -------
+        float
+            average function execution time
 
-        n_max_slow: integer
-            maximum n run for the "slow" transform methods, so far including 
-            only the "direct_python" implementation.
+        Notes
+        -----
+        The measurements overhead can be estimated by executing ::
+
+            Timent(...).time(lambda: None)
+
+        with a sufficiently large number of iterations (to avoid rounding
+        errors due to the finite timer precision).
+        In 2018, this overhead was on the order of 100 ns per iteration.
```

Comment:
It's 2019 now :)

Reply:
It still _is_. I'm not a prophet to write “In 2019 ... was”.  ;–)
And, I think, none of us actually has a 2019 system (I have one from 2018 and one from 2015).